lib/tesla/middleware/path_params.ex

```diff
@@ -34,9 +54,50 @@ defmodule Tesla.Middleware.PathParams do
 
   defp build_url(url, nil), do: url
 
+  defp build_url(url, %_{} = params), do: build_url(url, Map.from_struct(params))
+
   defp build_url(url, params) do
-    Regex.replace(@rx, url, fn match, key ->
-      to_string(params[String.to_existing_atom(key)] || match)
-    end)
+    params
+    |> Enum.map(fn {name, value} -> {to_string(name), value} end)
+    |> Enum.sort_by(fn {name, _} -> String.length(name) end, :desc)
+    |> Enum.reduce(url, &replace_parameter/2)
   end
+
+  # Do not replace parameters with nil values.
+  defp replace_parameter({_name, nil}, url), do: url
+  defp replace_parameter({name, value}, url), do: replace_parameter(to_string(name), value, url)
+
+  defp replace_parameter(<<h, t::binary>>, value, url)
+       when h in ?a..?z or h in ?A..?Z do
```

Comment:
I wonder if we should care at all in the end 🤔

Reply:
This was put in part because of the previous restrictions. For phoenix-style parameters, we absolutely care, IMO. For OpenAPI parameters…less so, but I still think that the behaviours documented are good.